### DISCUSSION TITLE
Show plot until next update

### DISCUSSION THREAD

**Discussion by shloksobti** [alandefreitas/matplotplusplus#131] [Q&A] [answered]
Hi! 
Is there a method to show a plot until it's updated? I don't want to have to press the "enter" key over again to update.

**Comment by alandefreitas** [comment]
If you're calling the drawing functions in a loop, just remove `matplot::show()` or `ax->show()` from your example. Or only call `show()` after the loop.

**Reply by alandefreitas** [reply]
An example: https://github.com/alandefreitas/matplotplusplus/blob/master/examples/geography/geoplot/geoplot_6.cpp

**Reply by shloksobti** [reply]
Thanks for the reply! I'm a little confused - removing [show()] doesn't show the plot, i.e. it disappears immediately. I want it to show until I update the plot.

**Reply by alandefreitas** [reply]
`draw()` and functions that call `draw()` only draw the plot. The `show()` function keeps the plot alive until there's some form of user interaction.

In more interactive applications, you usually keep updating the plot with `draw()` and then call `show()` only once in the end so that the program doesn't end and the plot doesn't disappear.

If you have a minimal example to replicate your problem, it might be easier to adapt that.

**Comment by shloksobti** [comment, marked as answer]
Thanks! I guess, I want some sort of a "live plotting". My application is plotting beliefs and I want to visualize how these beliefs change in an online inference setting.

If I `show()` only once at the end, it basically just shows the final belief. However, if I `show()` each time, it waits for the user to press Enter. One thing I can do is do all the plotting in a separate thread, but then I'd still need to somehow bypass the "press Enter to continue" step. Any thoughts?

**Reply by shloksobti** [reply]
For anyone else interested, just call `draw()` and then `sleep()` until you want to update the plot! Cheers!

**Reply by alandefreitas** [reply]
> If I show() only once at the end

> if I show() each time

Don't do any of these things. Don't only call `show` at the end. Don't call `show` each time.

Call `draw` while the program is running so you don't need user input AND call `show` at the end if you want the final results to be visible until there's some input.

These two functions serve different purposes. `draw` means "draw", but `show` means "draw and wait".

> One thing I can do is do all the plotting in a separate thread,

The plots use a process pipe, so they are *already* happening in a separate thread. Just keep calling `draw` and let your executable run as usual. 

Your plots will be visible in another window. You don't need to call `sleep`.

> bypass the "press Enter to continue" step

That's exactly what `draw()` does
